File: Low Level Design (LLD)/Problems/05. FoodDeliverySystem/Java/Main.java
import models.*;
import payment.Payment;
import service.*;
import utils.*;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Main {
    public static void main(String[] args) {

        // Initialize services
        UserService userService = new UserService();
        RestaurantService restaurantService = new RestaurantService();
        OrderService orderService = new OrderService();
        PaymentService paymentService = new PaymentService();
        DeliveryService deliveryService = new DeliveryService();
        CouponService couponService = new CouponService();
        NotificationService notificationService = new NotificationService();

        // Create User
        User customer = new User(IDGenerator.generateID(), "Alice", "alice@example.com", "9821857874", UserRole.CUSTOMER);
        userService.registerUser(customer);

        // Create Food Items
        FoodItem pizza = new FoodItem(IDGenerator.generateID(), "Chicken Pizza", 100.0);
        FoodItem burger = new FoodItem(IDGenerator.generateID(), "Veg Burger", 70.0);
        List<FoodItem> menu = Arrays.asList(pizza, burger);

        // Create a Restaurant
        Restaurant restaurant = new Restaurant(IDGenerator.generateID(), "Street-Za", "123 Main Street", menu);
        restaurantService.addRestaurant(restaurant);

        // Place an order
        List<FoodItem> itemsOrdered = Arrays.asList(pizza, burger);
        double totalAmount = itemsOrdered.stream().mapToDouble(FoodItem::getPrice).sum();
        Order order = new Order(IDGenerator.generateID(), customer, restaurant, itemsOrdered, totalAmount, OrderStatus.PLACED);
        orderService.placeOrder(order);
        notificationService.sendNotification("Restaurant is preparing your Order!", customer.getName());

        // Process Payment
        Payment payment = new Payment(IDGenerator.generateID(), order.getId(), order.getTotalAmount(), PaymentStatus.PENDING);
        paymentService.processPayment(payment);

        // Register a delivery person
        DeliveryPerson dp = new DeliveryPerson(IDGenerator.generateID(), "Bob", "9812341125", DeliveryStatus.AVAILABLE);
        deliveryService.registerDeliveryPerson(dp);

        // Assign Delivery
        DeliveryPerson assigned = deliveryService.getAvailableDeliveryPerson();
        if (assigned != null) {
            assigned.setStatus(DeliveryStatus.ON_DELIVERY);
            order.setStatus(OrderStatus.OUT_FOR_DELIVERY);
            notificationService.sendNotification("Your order is out for delivery!", customer.getName());
        }

        // Order Delivered
        order.setStatus(OrderStatus.DELIVERED);
        assigned.setStatus(DeliveryStatus.AVAILABLE);
        notificationService.sendNotification("Order delivered Successfully!", customer.getName());

        System.out.println("Hurray!! Order completed!");
    }
}

File: Low Level Design (LLD)/Problems/05. FoodDeliverySystem/Cpp/main.cpp
#include "./models/include/DeliveryPerson.h"
#include "./models/include/FoodItem.h"
#include "./models/include/Location.h"
#include "./models/include/Order.h"
#include "./models/include/Restaurant.h"
#include "./models/include/User.h"
#include "./payment/include/Coupon.h"
#include "./payment/include/Payment.h"
#include "./service/include/CouponService.h"
#include "./service/include/DeliveryService.h"
#include "./service/include/NotificationService.h"
#include "./service/include/OrderService.h"
#include "./service/include/PaymentService.h"
#include "./service/include/RestaurantService.h"
#include "./service/include/UserService.h"
#include "./utils/include/IDGenerator.h"

int main() {
    UserService userService;
    RestaurantService restaurantService;
    OrderService orderService;
    PaymentService paymentService;
    DeliveryService deliveryService;
    NotificationService notificationService;

    // Create User
    std::shared_ptr<User> customer = std::make_shared<User>(IDGenerator::generateID(), "Alice", "alice@example.com", "9821857874", UserRole::CUSTOMER);
    userService.registerUser(customer);

    // Create Food Items
    FoodItem pizza(IDGenerator::generateID(), "Chicken Pizza", 100.0);
    FoodItem burger(IDGenerator::generateID(), "Veg Burger", 70.0);
    std::vector<FoodItem> menu = {pizza, burger};

    // Create a Restaurant
    std::shared_ptr<Restaurant> restaurant = std::make_shared<Restaurant>(IDGenerator::generateID(), "Street-Za", "123 Main Street", menu);
    restaurantService.addRestaurant(restaurant);

    // Place an Order
    std::vector<FoodItem> itemsOrdered = {pizza, burger};
    double totalAmount = pizza.getPrice() + burger.getPrice();
    std::shared_ptr<Order> order = std::make_shared<Order>(IDGenerator::generateID(), *customer, *restaurant, itemsOrdered, totalAmount, OrderStatus::PLACED);
    orderService.placeOrder(order);
    notificationService.sendNotification("Restaurant is preparing your Order!", customer->getName());

    // Process Payment
    Payment payment(IDGenerator::generateID(), order->getId(), order->getTotalAmount(), PaymentStatus::PENDING);
    paymentService.processPayment(payment);

    // Register and Assign Delivery
    std::shared_ptr<DeliveryPerson> deliveryPerson = std::make_shared<DeliveryPerson>(IDGenerator::generateID(), "Bob", "9812341125", DeliveryStatus::AVAILABLE);
    deliveryService.registerDeliveryPerson(deliveryPerson);

    std::shared_ptr<DeliveryPerson> assigned = deliveryService.getAvailableDeliveryPerson();
    if (assigned != nullptr) {
        assigned->setStatus(DeliveryStatus::ON_DELIVERY);
        order->setStatus(OrderStatus::OUT_FOR_DELIVERY);
        notificationService.sendNotification("Your order is out for delivery!", customer->getName());
    }

    // Order Delivered
    order->setStatus(OrderStatus::DELIVERED);
    assigned->setStatus(DeliveryStatus::AVAILABLE);
    notificationService.sendNotification("Order delivered Successfully!", customer->getName());

    std::cout << "Hurray!! Order completed!" << std::endl;

    return 0;
}
